staging/src/k8s.io/apiserver/pkg/admission/plugin/cel/validator.go

```diff
@@ -0,0 +1,284 @@
+}
+
+// Compile compiles the cel expression defined in ValidatingAdmissionPolicy
+func (c *CELValidatorCompiler) Compile(p *v1alpha1.ValidatingAdmissionPolicy) (Validator, error) {
+	if len(p.Spec.Validations) == 0 {
+		return nil, nil
+	}
+	hasParam := false
+	if p.Spec.ParamKind != nil {
+		hasParam = true
+	}
+	compilationResults := make([]CompilationResult, len(p.Spec.Validations))
+	for i, validation := range p.Spec.Validations {
+		compilationResults[i] = CompileValidatingPolicyExpression(validation.Expression, hasParam)
+	}
+	return &CELValidator{policy: p, compilationResults: compilationResults}, nil
+}
+
+// CELValidator implements the Validator interface
+type CELValidator struct {
+	policy             *v1alpha1.ValidatingAdmissionPolicy
+	compilationResults []CompilationResult
+}
+
+func convertObjectToUnstructured(obj interface{}) *unstructured.Unstructured {
+	var unstructuredObj unstructured.Unstructured
+	jsonObj, _ := json.Marshal(obj)
+	_ = json.Unmarshal(jsonObj, &unstructuredObj.Object)
+
+	return &unstructuredObj
+}
+
+// Validate validates all cel expressions in Validator and returns PolicyDecisions and error.
+func (v *CELValidator) Validate(a admission.Attributes, o admission.ObjectInterfaces, params *unstructured.Unstructured) ([]PolicyDecision, error) {
+	decisions := make([]PolicyDecision, len(v.compilationResults))
+	var oldObjectVal *unstructured.Unstructured
+	if a.GetOldObject() != nil {
+		oldObjectVal = convertObjectToUnstructured(a.GetOldObject())
+	}
+
+	var objectVal *unstructured.Unstructured
+	if a.GetObject() != nil {
+		objectVal = convertObjectToUnstructured(a.GetObject())
+	}
+	va := &validationActivation{
+		object:    objectVal,
+		oldObject: oldObjectVal,
+		params:    params,
+		request:   convertObjectToUnstructured(createAdmissionRequest(a)),
+	}
+
+	for i, compilationResult := range v.compilationResults {
+		var policyDecision = &decisions[i]
+
+		if compilationResult.Error != nil {
+			f := *v.policy.Spec.FailurePolicy
+			if f == v1alpha1.Ignore {
+				policyDecision.Kind = Admit
+			} else {
+				policyDecision.Kind = Deny
+			}
+			policyDecision.Message = fmt.Sprintf("compilation error: %v", compilationResult.Error)
+			continue
+		}
+		if compilationResult.Program == nil {
+			f := *v.policy.Spec.FailurePolicy
+			if f == v1alpha1.Ignore {
+				policyDecision.Kind = Admit
+			} else {
+				policyDecision.Kind = Deny
+			}
+			policyDecision.Message = fmt.Sprintf("compilation result is nil")
+			continue
+		}
+		evalResult, _, err := compilationResult.Program.Eval(va)
+		if err != nil {
+			f := *v.policy.Spec.FailurePolicy
+			if f == v1alpha1.Ignore {
+				policyDecision.Kind = Admit
+			} else {
+				policyDecision.Kind = Deny
+			}
+			policyDecision.Message = fmt.Sprintf("failed expression: %v, with evaluation error: %v", v.policy.Spec.Validations[i].Expression, err)
+		} else if evalResult != celtypes.True {
+			policyDecision.Kind = Deny
+			reason := v.policy.Spec.Validations[i].Reason
```

Comment:
I missed this index call before. Add an assertion somewhere that the length of `v.policy.Spec.Validations` matches the length of `v.compilationResults`?

Reply:
The compilationResults generated in above compile func which initialized to have the same length as `v.policy.Spec.Validations`